### PR TITLE
Add Azure IoT Hub

### PR DESCRIPTION
**Pull request by ahryshchanka**

### CODE REVIEW

**Comment by ahryshchanka** [comment]
issue #335

File: src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+using System;
+
+namespace HealthChecks.AzureIoTHub
+{
+    public class AzureIoTHubOptions
+    {
+        internal string ConnectionString { get; }
+        internal bool RegistryReadCheck { get; private set; } = false;
```
Comment (sungam3r) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs]:
```suggestion
        internal bool RegistryReadCheck { get; private set; }
```

File: src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        internal bool RegistryWriteCheck { get; private set; } = false;
```
Comment (sungam3r) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs]:
```suggestion
        internal bool RegistryWriteCheck { get; private set; }
```

File: src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        internal bool ServiceConnectionCheck { get; private set; } = false;
```
Comment (sungam3r) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs]:
```suggestion
        internal bool ServiceConnectionCheck { get; private set; }
```

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                {
+                    using (var client = RegistryManager.CreateFromConnectionString(_options.ConnectionString))
+                    {
+                        var deviceId = _options.RegistryWriteDeviceIdFactory();
+
+                        await client.AddDeviceAsync(new Device(deviceId), cancellationToken);
+                        await client.RemoveDeviceAsync(deviceId, cancellationToken);
+                    }
+                }
+                else if (_options.RegistryReadCheck)
```
Comment (sungam3r) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: Are `RegistryWriteCheck` and `RegistryReadCheck` mutually exclusive options?

Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: Yes

you can write => you can read (always)
you can read => possibly you can write
read
read-write

File: src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs (hunk 1 above)
Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs]: fixed

File: src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs (hunk 3 above)
Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs]: fixed

File: src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs (hunk 2 above)
Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubOptions.cs]: fixed

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            try
+            {
+                if (_options.RegistryWriteCheck)
+                {
+                    using (var client = RegistryManager.CreateFromConnectionString(_options.ConnectionString))
+                    {
+                        var deviceId = _options.RegistryWriteDeviceIdFactory();
+
+                        await client.AddDeviceAsync(new Device(deviceId), cancellationToken);
```
Comment (sungam3r) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: What if device with that id already exists?

Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: Exception will throw, I think, and unhealthy result will be. I'll check this scenario.

Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: Reimplemented:
if device with id already exists -- remove it
if device with id not exists -- create and remove it

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            var transportType = MapToTransportType(_options.ServiceConnectionTransport);
+            using (var client = ServiceClient.CreateFromConnectionString(_options.ConnectionString, transportType))
+            {
+                var result = await client.GetServiceStatisticsAsync(cancellationToken);
+                var _ = result.ConnectedDeviceCount;
+            }
+        }
+
+        private async Task ExecuteRegistryReadCheck()
```
Comment (sungam3r) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: I would prefer Async suffix here and above.

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+
+        private async Task ExecuteServiceConnectionCheckAsync(CancellationToken cancellationToken)
+        {
+            var transportType = MapToTransportType(_options.ServiceConnectionTransport);
+            using (var client = ServiceClient.CreateFromConnectionString(_options.ConnectionString, transportType))
+            {
+                var result = await client.GetServiceStatisticsAsync(cancellationToken);
+                var _ = result.ConnectedDeviceCount;
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: remove var, discard does not need it

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                return new HealthCheckResult(context.Registration.FailureStatus, exception: ex);
+            }
+        }
+
+        private async Task ExecuteServiceConnectionCheckAsync(CancellationToken cancellationToken)
+        {
+            var transportType = MapToTransportType(_options.ServiceConnectionTransport);
+            using (var client = ServiceClient.CreateFromConnectionString(_options.ConnectionString, transportType))
+            {
+                var result = await client.GetServiceStatisticsAsync(cancellationToken);
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: rename result as statistics

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                var _ = result.ConnectedDeviceCount;
+            }
+        }
+
+        private async Task ExecuteRegistryReadCheckAsync()
+        {
+            using (var client = RegistryManager.CreateFromConnectionString(_options.ConnectionString))
+            {
+                var query = client.CreateQuery(_options.RegistryReadQuery, 1);
+                var _ = await query.GetNextAsJsonAsync();
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: remove var, discard does not need it

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+
+        /// <summary>
+        /// This method try create and remove device for registry write check.
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: can you add explanation inside methdo, typically we don't add summary comments to private members

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs (hunk 8 above)
Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: variable were removed as unused, enough to execute method (to do required) without any manipulations with `result/statistics`

File: src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs (hunk 10 above)
Comment (ahryshchanka) [reply src/HealthChecks.AzureIoTHub/AzureIoTHubtHealthCheck.cs]: It is duplicate for already added in method comment. Removed

**Comment by unaizorrilla** [comment]
Look good, y try to merge  after my working day!

**Comment by ahryshchanka** [comment]
@unaizorrilla 
does it make sense to change naming for package based on issue #342 ?

**Comment by unaizorrilla** [comment]
Hi @ahryshchanka 

First of all, I'm so sorry for my delay reviewing and merging this PR. Now is merged with some minor changes!

Thanks again for contribute on this project creating a new healtcheck for Azure IoT.

Build is on app veyor, when finish a new release build will create the package and upload to NuGet.

**Comment by ahryshchanka** [comment]
Hi @unaizorrilla .

Thank you for your excellent library! i'm glad to help with it.

**Pull request merged**